core/trino-main/src/main/java/io/trino/operator/aggregation/DistinctWindowAccumulator.java

```diff
@@ -0,0 +1,195 @@
+    private final List<Integer> argumentChannels;
+    private final MarkDistinctHash hash;
+    private final PageBuilder pageBuilder;
+    private final PagesIndex.Factory pagesIndexFactory;
+
+    public DistinctWindowAccumulator(
+            WindowAccumulator delegate,
+            List<Type> argumentTypes,
+            List<Integer> argumentChannels,
+            FlatHashStrategyCompiler hashStrategyCompiler,
+            Session session,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.argumentTypes = ImmutableList.copyOf(argumentTypes);
+        this.argumentChannels = ImmutableList.copyOf(argumentChannels);
+        this.hash = new MarkDistinctHash(
+                session,
+                argumentTypes,
+                false,
+                hashStrategyCompiler,
+                UpdateMemory.NOOP);
+
+        this.pageBuilder = new PageBuilder(argumentTypes);
+        this.pagesIndexFactory = requireNonNull(pagesIndexFactory, "pagesIndexFactory is null");
+    }
+
+    private DistinctWindowAccumulator(
+            WindowAccumulator delegate,
+            List<Type> argumentTypes,
+            List<Integer> argumentChannels,
+            MarkDistinctHash hash,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.delegate = delegate;
+        this.argumentTypes = argumentTypes;
+        this.argumentChannels = argumentChannels;
+        this.hash = hash;
+        this.pageBuilder = new PageBuilder(argumentTypes);
+        this.pagesIndexFactory = pagesIndexFactory;
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return delegate.getEstimatedSize()
+                + pageBuilder.getRetainedSizeInBytes()
+                + hash.getEstimatedSize();
+    }
+
+    @Override
+    public WindowAccumulator copy()
+    {
+        indexCurrentPage();
+        return new DistinctWindowAccumulator(
+                delegate.copy(),
+                argumentTypes,
+                argumentChannels,
+                hash.copy(),
+                pagesIndexFactory);
+    }
+
+    @Override
+    public void addInput(WindowIndex index, int startPosition, int endPosition)
+    {
+        // index is remapped so just go from 0 to argumentChannels.size()
+        for (int position = startPosition; position <= endPosition; position++) {
+            if (pageBuilder.isFull()) {
+                indexCurrentPage();
+            }
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                ValueBlock value = index.getSingleValueBlock(channel, position).getSingleValueBlock(0);
+                pageBuilder.getBlockBuilder(channel).append(value, 0);
+            }
+            pageBuilder.declarePosition();
+        }
+    }
+
+    private void indexCurrentPage()
+    {
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
```

Comment:
Could we build the Page and reset the builder in the caller?

Reply:
If you wish :)